oak/server/channel.h

```diff
@@ -23,123 +23,157 @@
 };
 
 // Shared-ownership wrapper for the read half of a MessageChannel.
-class MessageChannelReadHalf final : public ChannelHalf {
+class MessageChannelReadHalf {
  public:
   MessageChannelReadHalf(std::shared_ptr<MessageChannel> channel) : channel_(channel) {}
-  ReadResult Read(uint32_t size) override { return channel_->Read(size); }
-  bool CanRead() override { return channel_->Count() > 0; }
-  ReadResult BlockingRead(uint32_t size) override { return channel_->BlockingRead(size); }
-  void Await() override { return channel_->Await(); }
+
+  std::unique_ptr<MessageChannelReadHalf> Clone() {
+    return absl::make_unique<MessageChannelReadHalf>(channel_);
+  }
+  // Read a message of up to |size| bytes from the Channel. The caller owns any
+  // returned message, whose actual size of may be less than |size|.  If the
+  // next available message on the channel is larger than |size|, no data will
+  // be returned and the required_size field of the result will indicate the
+  // required size.
+  ReadResult Read(uint32_t max_size, uint32_t max_channels) {
+    return channel_->Read(max_size, max_channels);
+  }
+
+  // BlockingRead behaves like Read but blocks until a message is available.
+  ReadResult BlockingRead(uint32_t max_size, uint32_t max_channels) {
+    return channel_->BlockingRead(max_size, max_channels);
+  }
+
+  // Indicate whether a Read operation would return a message.
+  bool CanRead() { return channel_->Count() > 0; }
+
+  // Await blocks until there is a message available to read on the channel.
+  void Await() { return channel_->Await(); }
 
  private:
   std::shared_ptr<MessageChannel> channel_;
 };
 
 // Shared-ownership wrapper for the write half of a MessageChannel.
-class MessageChannelWriteHalf final : public ChannelHalf {
+class MessageChannelWriteHalf {
  public:
   MessageChannelWriteHalf(std::shared_ptr<MessageChannel> channel) : channel_(channel) {}
-  void Write(std::unique_ptr<Message> msg) override { channel_->Write(std::move(msg)); }
+
+  std::unique_ptr<MessageChannelWriteHalf> Clone() {
+    return absl::make_unique<MessageChannelWriteHalf>(channel_);
+  }
+
+  // Write the provided message to the Channel.
+  void Write(std::unique_ptr<Message> msg) { channel_->Write(std::move(msg)); }
 
  private:
   std::shared_ptr<MessageChannel> channel_;
 };
 
+// Owning reference to either a read or write half of a channel.
+// TODO: Use std::variant when C++17 available.
+struct ChannelHalf {
+  ChannelHalf(std::unique_ptr<MessageChannelReadHalf> r) : which(Half::READ), read(std::move(r)) {}
+  ChannelHalf(std::unique_ptr<MessageChannelWriteHalf> w)
+      : which(Half::WRITE), write(std::move(w)) {}
+
+  std::unique_ptr<ChannelHalf> Clone() {
+    switch (which) {
+      case Half::READ:
+        return absl::make_unique<ChannelHalf>(read->Clone());
+        break;
+      case Half::WRITE:
+        return absl::make_unique<ChannelHalf>(write->Clone());
+        break;
```

Comment:
Can we add a default option and print an error message?

Reply:
(Overtaken by move to `absl::variant`)